Include elements with pointer cursor style as clickable links.

tylercal opened the pull request:
The pointer cursor is a UI hint that something can be clicked on the page. This is often the simplest way to detect if a non-input, non-anchor element should have a hint. The "My Calendars" and "Other calendars" sections of Google Calendar are one example.

This change uses `getComputedStyle` to determine the value of an element's cursor. We only want to know the first element that sets its cursor to pointer. All children of that first element will also inherit the style and because this method includes inherited styles we need a special check to make sure a given style is in fact the first pointer in the tree.

gdh1995 commented:
But some pages may set `cursor: pointer` on `<html>` ... so I think we need more conditions for this rule.

BTW, as far as I know, it's hard to know if one node's `cursor` property is `inherit`.

gdh1995 commented:
(I'm sorry I gave an useless comment and I've deleted it.)

This seems a little too costly to compute `getComputedStyle` twice on every node.

smblott-github commented:
> This seems a little too costly to compute getComputedStyle twice on every node.

Yes, that was my concern too.  we should measure the cost.

tylercal commented:
Thanks for the feedback @gdh1995, good idea @smblott-github.

These results were gathered by looping 10 times with each strategy. So the average performance hit per loop is the difference divided by 10.

| URL                                         | w/o   | with  | avg hit / loop (ms) |
|---------------------------------------------|-------|-------|----------|
| https://www.google.com/search?q=vimium      | 218.3 | 332.3 | 11.4     |
| https://github.com/philc/vimium             | 204.6 | 234.7 | 3.01     |
| https://www.reddit.com/r/chrome/            | 559.7 | 626.2 | 6.65     |
| https://calendar.google.com/calendar/render | 265.8 | 339.4 | 7.36     |

Specifically, I added a switch to turn the feature on and off and set up the loop as follows:

```coffeescript
    console.profile("getVisibleClickable")
    for x in [0..10]
      visibleElements = []
      for element in elements
        unless requireHref and not element.href
          visibleElement = @getVisibleClickable(element, false)
          visibleElements.push visibleElement...
    console.profileEnd("getVisibleClickable")

    console.profile("getVisibleClickable-pointer")
    for x in [0..10]
      visibleElements = []
      for element in elements
        unless requireHref and not element.href
          visibleElement = @getVisibleClickable(element, true)
          visibleElements.push visibleElement...
    console.profileEnd("getVisibleClickable-pointer")
```

I can explore some strategies to improve the performance further, but it will likely come at the cost of readability and/or memory use.

tylercal commented:
I switched up how I was measuring, used the median of 11 tries, and added an optimized version that checks later in the methods and filters elements that have no class attribute.

In most cases the hit is less than 30ms with both versions. I think it is worth the extra time in all cases to not have to reach for the mouse.

| URL | Old ms | New ms | Opt ms | New % chg | Opt % chg |
|----------------------------------------------------|--------|--------|--------|---------|---------|
| https://www.google.com/search?q=vimium | 42.95 | 72.37 | 65.06 | 68.49% | 51.49% |
| https://github.com/philc/vimium | 45.23 | 62.76 | 57.00 | 38.76% | 26.05% |
| https://www.reddit.com/r/chrome/ | 98.72 | 184.25 | 117.52 | 86.65% | 19.05% |
| https://calendar.google.com/calendar/render#main_7 | 68.91 | 62.61 | 52.39 | -9.14% | -23.97% |
| https://www.facebook.com/ | 148.27 | 256.54 | 179.12 | 73.03% | 20.81% |
| https://en.wikipedia.org/wiki/Main_Page | 45.13 | 66.87 | 53.35 | 48.17% | 18.21% |
| https://www.yahoo.com/ | 41.40 | 56.34 | 48.51 | 36.10% | 17.17% |
| https://www.amazon.com/ | 116.95 | 258.25 | 145.45 | 120.81% | 24.36% |

New testing loop:
```coffee
    g1 = []
    g2 = []
    g3 = []

    timer1 = 0
    timer2 = 0
    timer3 = 0

    for x in [0..10]
      visibleElements = []
      for element in elements
        unless requireHref and not element.href
          s = performance.now()
          visibleElement = @getVisibleClickable(element, 0)
          timer1 += performance.now() - s
          visibleElements.push visibleElement...

      visibleElements = []
      for element in elements
        unless requireHref and not element.href
          s = performance.now()
          visibleElement = @getVisibleClickable(element, 1)
          timer2 += performance.now() - s
          visibleElements.push visibleElement...

      visibleElements = []
      for element in elements
        unless requireHref and not element.href
          s = performance.now()
          visibleElement = @getVisibleClickable(element, 2)
          timer3 += performance.now() - s
          visibleElements.push visibleElement...

      g1.push timer1
      g2.push timer2
      g3.push timer3
```

Optimized check (after has button class check):
```coffee
    if not isClickable and
      element.className != "" and
      document.defaultView.getComputedStyle(element, null).cursor == "pointer" and
      document.defaultView.getComputedStyle(element.parentElement, null).cursor != "pointer"
        isClickable = true
```

Let me know if I should update the PR with the "optimized" version.

gdh1995 commented:
We may need to test a much bigger&complicated page which contains tons of nodes in order to know the worst performance. Any ideas?

I can not visit Google or Facebook websites now, so I know little about them. But will a long news feed page be suitable?

smblott-github commented:
A good worst-case test is to go to Twitter, then scroll down and down and down.... and so on, and *then* try link hints.

(Regardless of this PR, Vimium has performance issues in this scenario.)

tylercal commented:
Performed the tests on a slightly faster computer. I'm guessing the cases that the original was slower are mostly noise since I didn't get that result on my laptop. The worst case results still look pretty reasonable for what you gain.

"Nodes" is the number of elements visited by a single loop of all elements.

| URL | Nodes | Old | New | Opt | New | Opt |
|----------------------------------------------------------|----------|-----|------|------|------|------|
| https://www.google.com/search?q=vimium | 876 | 44 | 59 | 38 | 33% | -14% |
| https://github.com/philc/vimium | 1307 | 50 | 63 | 40 | 26% | -19% |
| https://www.reddit.com/r/news/ | 2510 | 79 | 122 | 97 | 54% | 23% |
| https://calendar.google.com/calendar/render?pli=1#main_7 | 882 | 44 | 59 | 42 | 32% | -7% |
| https://www.facebook.com/ | 1861 | 88 | 130 | 78 | 47% | -12% |
| https://en.wikipedia.org/wiki/Main_Page | 959 | 57 | 48 | 45 | -16% | -21% |
| https://www.yahoo.com/ | 916 | 56 | 58 | 47 | 4% | -17% |
| https://www.amazon.com/ | 3600 | 108 | 235 | 134 | 117% | 24% |
| https://twitter.com/ | 5320 | 163 | 329 | 197 | 102% | 21% |
| https://twitter.com/ (scrolled 10 pages - 230 tweets) | 30009 | 804 | 1878 | 1316 | 133% | 64% |


An "interesting" way to handle the pages with many elements would be to try [elementFromPoint](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint). That might have it's own interesting performance problems though - fodder for another PR.

tylercal commented:
Are there any other tests you'd like to try? What are the next steps for this PR?

gdh1995 commented:
It's indeed effective. Suggest to merge this if we can not merge #1859 `Enable link hints for all elements with `click` listeners`

smblott-github commented:
Thanks for the detailed performance numbers, @tylercal.  Unfortunately, I'd say those numbers aren't encouraging.

Is there any other way to get at the elements we need?

tylercal commented:
I've been using Vimium with this change for over a week in all my real world browsing and haven't noticed any performance impact. Of course with the performance hit in the 10's of milliseconds, I wouldn't expect to be able to detect a difference. 

@smblott-github - could you offer some guidance on what kind of numbers would be encouraging?

If it's just that you don't want to make the 10,000+ elements case any worse, a decent compromise would be to disable the check when there are more than 10k or 20k nodes. I personally don't mind the wait since a 500ms extra delay is still faster than reaching for the mouse.

I'm happy to keep experimenting, but it will be helpful to know what we're aiming for.

smblott-github commented:
@tylercal... I think the status quo (without this PR) is too slow, and I'm not keen on adding anything that makes it slower.  I'd rather look into a way of speeding up what we have, first.

tylercal commented:
Thanks for the extra clarity.

No matter what kinds of optimizations we come up with, there is always going to be some worse case page that breaks the optimization. For example, a page that has 500k visible elements by just nesting divs forever. 

It will of course take longer than this change, but I'll see what I can come up with speed improvement-wise in another PR. Are there any open issues you can point too with good test pages that users have reported trouble with? I can at least start there.

eejdoowad commented:
Can this be a configurable option? It really improves the user experience on some sites (e.g. github line numbers) and some computers are fast enough that the performance hit doesn't really matter.

narve commented:
Yes, please merge this one, at least as a configurable option! Vimium is so addictive that I hate not being able to click on everything :)

smblott-github commented:
There might be legs in this idea yet, @tylercal.

For example, consider this...

```Coffeescript
   if not isClickable and
      element.className != "" and
      document.defaultView.getComputedStyle(element, null).cursor == "pointer" and
      document.defaultView.getComputedStyle(element.parentElement, null).cursor != "pointer"
        isClickable = true
```

tylercal commented:
@smblott-github I've been researching ways of computing only for what is displayed on the screen to speed up finding links to hint, but nothing is as "fast" as what we currently have. Unfortunately most are more than 1000x slower.

Checking that the element first has a class should help, but there will always be contrived examples that break these kinds of optimizations (e.g. bootstrap pages will have more elements with classes than not). I think I shared this sentiment in my last comment and the issue still remains. The open questions are **what amount of optimization is enough? And on what set of pages?**

Your last answer was that it's already too slow, and we'd have to speed up what exists today first. If we're ready to take a reasonable performance hit, I'll start toying around with the optimizations again.

For what it's worth, I've been using my forked version of Vimium with this change for a year now. Performance still doesn't feel any different than "vanilla" Vimium in day to day browsing.

smblott-github commented:
> @smblott-github I've been researching ways of computing only for what is displayed on the screen to speed up finding links to hint, but nothing is as "fast" as what we currently have. Unfortunately most are more than 1000x slower.

Yes.  I went through a similar process myself some time back.

I wrote this previously...

> Because we visit parents before children, we might be able to avoid one of those .getComputedStyle() calls.

I tried that idea out.  For the very first page I tested it on ([Feedly](https://feedly.com/)), it was selecting the wrong clickable elements (actually, ancestors of the clickable elements).  So I stopped thinking about it.

I guess your experience has been more positive?  In light of #2884, we might need to disable #2871, in which case the approach here might be an attractive alternative.
